### DISCUSSION TITLE
Is disabling/enabling VL53L1X via I2CDriver54 0/1 command supposed to trigger reboot (ESP32)?

### DISCUSSION THREAD

**Discussion by dr-knor** [arendst/Tasmota#21105] [Q&A] [answered]
I am trying to use a VL53L1X sensor in a project that only requires a readout every 5 minutes or so. To save power, my (naive?) idea was to enable/disable the sensor when i want to do a measurement (and use the XSHUT line to turn it off completely). I noticed, however that the ESP32 reboots when I give an I2CDriver (0 or 1) command.   Is this expected behavior? The problem with that is that my Berry script/driver also stops working when that happens. The documentation speaks about using the I2CDriver command to enable disable sensors at runtime, but is the reboot avoidable?

**Comment by sfromis** [comment, marked as answer]
Yes, switching I2CDriver is supposed to trigger a reboot, as this is where the sensor driver becomes enabled or not. The energy consumption of the sensor is so low that any extra steps you do to try to mitigate his is likely to cost more power than the potential saving, even without reboots.

If you want to only process readings at TelePeriod, you can use triggers for that, in Berry this means prefixing the trigger with `tele#`.

Rebooting should not stop Berry code from running (assuming that you've set it up to start again), but what happens here depends on details of what you mean.

**Reply by dr-knor** [reply]
Thanks. That clarifies it. I read in the STM documentation that the sensor can be operated in an ultra low power mode, but I think Tasmota uses it in continuous mode (one or more rangings per second it seems). I actually have no clue about what power saving ULP could bring and if I need it for my battery-operated use case. I will try to get a better understanding of that. If power consumption is OK without additional measures that would be nice of course.

**Reply by sfromis** [reply]
Tasmota is running on Espressif chips which are **really** not suitable for battery operation. The power draw needed to keep ESP8266 or ESP32 running (especially with Wifi) is what makes the VL53L1X consumption of marginal importance.

**Reply by dr-knor** [reply]
Hey, it is for a fun tinker project, no lives depend on it ;-). The battery will be charged by a solar panel (with a load sharing circuit between battery/panel/ESP). I will use an ESP32 board with favorable consumption characteristics (LolinD32) which I expect to run at 3-20 mA in modem sleep mode, maybe look at light/deep sleep if needed. The VL53L1X  datasheet mentions 20 uA for intermeasurement period (50 ms Tasmota setting) and 18 mA for ranging (140 ms Tasmota setting). This amounts to 12 mA continuous current draw, so indeed considerable relative to the ESP32 as you mention. If I would turn the sensor off for 98% of the time (on a 5 minute cycle), I could reduce the average sensor consumption by 60x to 200 uA. To enable the sensor I would have to trigger a reboot per your explanation above, briefly (?) resulting in a higher consumption after reboot when the ESP32 reconnects to Wifi. We'll see how it goes.

**Reply by sfromis** [reply]
Are you expecting Tasmota to have a feature for modem sleep mode?

Deep sleep is supported by Tasmota, but awaking from deep sleep is a full boot drawing notable power to complete all steps.

Tasmota also has support for [ESP32 ULP](https://tasmota.github.io/docs/ULP/), but that is outside of the normal sensor support, which is not operational in ULP mode.

**Reply by dr-knor** [reply]
Well, I had the impression Tasmota has some built-in power saving making use of modem sleep mode but I guess not then. Maybe deep sleep will be my best chance.
